nbextensions/bulkImportCell/paramsWidget.js

```diff
@@ -0,0 +1,551 @@
+define([
+    'bluebird',
+    'jquery',
+    // CDN
+    'kb_common/html',
+    // LOCAL
+    'common/ui',
+    'common/events',
+    'common/props',
+    'widgets/appWidgets2/fieldWidgetCompact',
+    'widgets/appWidgets2/paramResolver',
+    'common/runtime'
+], function (
+    Promise,
+    $,
+    html,
+    UI,
+    Events,
+    Props,
+    FieldWidget,
+    ParamResolver,
+    Runtime
+) {
+    'use strict';
+
+    let tag = html.tag,
+        form = tag('form'),
+        span = tag('span'),
+        div = tag('div');
+
+    function factory(config) {
+        let runtime = Runtime.make(),
+            paramsBus = config.bus,
+            workspaceInfo = config.workspaceInfo,
+            initialParams = config.initialParams,
+            container,
+            ui,
+            bus,
+            places = {},
+            model = Props.make(),
+            paramResolver = ParamResolver.make(),
+            settings = {
+                showAdvanced: null
+            },
+            widgets = [];
+
+
+        // DATA
+        /*
+       *
+       * Evaluates the parameter spec to determine which input widget needs to be
+       * invoked, but doesn't know what the widget does.
+       * Provides the communication bus for each input to route info to the widget
+       * and out of it.
+       *
+       *  In terms of widgets it looks like this:
+       *
+       *  InputCellWidget
+       *    inputWidgets
+       *      FieldWidget
+       *        textInputWidget
+       *      FieldWidget
+       *        objectInputWidget
+       *      FieldWidget
+       *        newObjectInputWidget
+       *      FieldWidget
+       *        integerInputWidget
+       *      FieldWidget
+       *        floatInputWidget
+       */
+
+        /*
+       * The input control widget is selected based on these parameters:
+       * - data type - (text, int, float, workspaceObject (ref, name)
+       * - input app - input, select
+       */
+
+
+        // RENDERING
+
+        /*
+      The field widget is a generic wrapper around the input. It serves the following purposes:
+      - intercepts messages in order to display status.
+      appSpec - specifies the whole app
+      parameterSpec - just the segment of the appSpec that specifies this individual parameter
+      value - the initial value of this field
+      closeParameters - a list of "close" parameters, which might be context-dependent. E.g. for an output
+          field, this would be the list of all parameters meant to be output objects, so their names can
+          be cross-validated for uniqueness (optional)
+      */
+        function makeFieldWidget(appSpec, parameterSpec, value, closeParameters) {
```

Comment:
field widgets seem like a good candidate to be moved further into /common, especially as they will be used for the file inputs as well.

Reply:
It already is! It lives in `widgets/appWidgets2`. Which maybe that directory name is a misnomer at this point.